pandas/tseries/offsets.py

```diff
@@ -275,14 +275,17 @@ def apply_index(self, i):
                        kwds.get('months', 0)) * self.n)
             if months:
                 shifted = liboffsets.shift_months(i.asi8, months)
-                i = i._shallow_copy(shifted)
+                i = type(i)(shifted, freq=i.freq, dtype=i.dtype)
 
             weeks = (kwds.get('weeks', 0)) * self.n
             if weeks:
                 # integer addition on PeriodIndex is deprecated,
                 #   so we directly use _time_shift instead
                 asper = i.to_period('W')
-                shifted = asper._data._time_shift(weeks)
+                if not isinstance(asper._data, np.ndarray):
```

Comment:
when is this not the case?

(or why was this change needed?)

Reply:
and something else, this full `if weeks` block does not seem to be covered (don't know if it is a false positive or not)

Reply:
`asper` at this point may be either a `PeriodIndex` or a `PeriodArray`.  If it is a `PeriodIndex` we want to unwrap it to get a PeriodArray.  If it is a PeriodArray, we _dont_ want to unwrap it since that would give a ndarray

Reply:
Good note on the `if weeks`.  Adding coverage is out of scope, but I'll add a note in the offsets TODO list